GowerMetric.py

```diff
@@ -154,16 +230,88 @@ def fit(self, X):
                     ratio_cols[:, zero_values_mask], axis=0
                 )
 
-            if self.ratio_scale_normalization == "kde":
-                n = X.shape[0]
-                c = 1.06
-
-                s = np.std(ratio_cols, axis=0)
-                self.h_ = (
-                    c / n ** (1 / 5) * np.min([s, self.ranges_ / 1.34], axis=0)
-                )
-
-        loader = GowerMetricWeights(self)
+            n = X.shape[0]
+
+            # h_t parameter
+            if self.ratio_scale_window == "kde":
+                if self.kde_type == "silverman":
+                    c = 1.06
+
+                    s = np.std(ratio_cols, axis=0)
+                    self.h_ = (
+                        c
+                        / n ** (1 / 5)
+                        * np.min([s, self.ranges_ / 1.34], axis=0)
+                    )
+                elif self.kde_type == "scott":
+                    c = 0.9
+
+                    s = np.std(ratio_cols, axis=0)
+                    self.h_ = c / n ** (1 / 5) * s
+                elif self.kde_type == "sheather-jones":
+                    from KDEpy import FFTKDE
+
+                    try:
+                        self.h_ = np.array(
+                            [
+                                FFTKDE(kernel="gaussian", bw="ISJ")
+                                .fit(ratio_cols[:, i])
+                                .bw
+                                for i in range(self.ratio_scale_num)
+                            ]
+                        )
+                    except ValueError:
+                        print(
+                            "FFTKDE resulted in failure - "
+                            "Crashed with error: ValueError: Root finding did not converge. Need more data. - "
+                            "Repeating fit using kde_type == 'silverman'"
+                        )
+
+                        # Try to fit using silverman instead
+                        self.kde_type = "silverman"
+                        self.fit(X)
+                        return
+                elif self.kde_type == "cv_grid":
+                    from sklearn.model_selection import GridSearchCV
+                    from sklearn.neighbors import KernelDensity
+
+                    self.h_ = np.array(
+                        [
+                            GridSearchCV(
+                                KernelDensity(),
+                                {"bandwidth": np.linspace(0.1, 1e06, 100)},
+                                cv=20,
+                            )
```

Comment:
1. You can create this object before loop.
2. Those are pretty weird values for bandwidth, aren't they? I think that log space is used more commonly. Why `1e06` instead of `1e6`? Why 20-fold CV, instead of more typical 10-fold, or LOOCV if you want an unbiased estimate? I would stick with 10-fold personally.

Reply:
Also change in Optuna CV